### PR TITLE
Benchmarking tools for FIT

### PR DESCRIPTION
**Pull request by mslupeck**
A set of tools to monitor the performance of O2. It works based on the memory consumption and CPU time of each process ID, that corresponds to the running FairMQ devices. Use is, in principle, not limited to FIT.

### CODE REVIEW

**Comment by mslupeck** [comment]
Hi @zsaldic 
Could you have a look at the details of the failed test and fix it?

**Comment by sawenzel** [comment]
Thanks. Note that these tools are based on already existing scripts used in various production tests. Probably it would be useful to add new features directly there.

File: Detectors/FIT/benchmark/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Documentation for Performance testing
+This document will summarize the tools that can be used to get information about the memory and CPU time evolution of simulations in ALICE O2. 
+
+In this folder you will find two scripts:
+1. `monitor.sh`
+2. `process.py`
+
+Both of these scripts define the two step procedure (**monitoring** and **processing**) of obtainging performance metrics of interest: _maximum memory, average memory, maximum CPU time, wall clock time, (CPU)/(wall clock) time ratio_ and lastly _plots of the evolving memory and cpu as a function of wall clock time._  
+
+
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/README.md]: This line has to be deleted

File: Detectors/FIT/benchmark/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+This document will summarize the tools that can be used to get information about the memory and CPU time evolution of simulations in ALICE O2. 
+
+In this folder you will find two scripts:
+1. `monitor.sh`
+2. `process.py`
+
+Both of these scripts define the two step procedure (**monitoring** and **processing**) of obtainging performance metrics of interest: _maximum memory, average memory, maximum CPU time, wall clock time, (CPU)/(wall clock) time ratio_ and lastly _plots of the evolving memory and cpu as a function of wall clock time._  
+
+
+### 1) Monitoring
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/README.md]: This line should be changed to: ## 1) Monitoring

File: Detectors/FIT/benchmark/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+### 1) Monitoring 
+You can monitor whatever you like as:
+
+`$> ./monitor.sh <your o2 command>`
+
+To obtain plots of (FairMQ) devices operating in the simulation you will have to generate a **logfile**  as:
+
+`$> ./monitor.sh <your o2 command> | tee o2xxx.log`
+
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/README.md]: This line should be deleted

File: Detectors/FIT/benchmark/process.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        pid_names = ['driver','event gen','sim worker 0','hit merger']
+        pid_vals = [driver,event_gen,sim_worker,hit_merger]
+
+        # append pid names for remaining workers
+        for i in range(n_workers-1):
+            pid_names.append(f"sim worker {i+1}")
+        
+        no_log = False
+    
+    except :
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/process.py]: This line should be changed to: except IOError:

File: Detectors/FIT/benchmark/process.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        pid_names.insert(0,'driver')
+        pid_vals.insert(0,driver)
+
+    #    for id in range(len(pid_names)):
+    #        print(pid_names[id],"PID is: ",pid_vals[id])
+    #        print(pid_vals[pid])
+    #    print("*******************************\n")
+        no_log = False
+        
+    except:
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/process.py]: This line should be changed to: except IOError:

File: Detectors/FIT/benchmark/process.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+            # save the first 100 lines in o2digi.log
+            loglines = [next(logfile) for line in range(100)]
+
+        # declare list for PID numbers and names
+        pid_vals = []
+        pid_names = []
+
+        # loop through lines to find PIDs
+        for line in range(len(loglines)):
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/process.py]: This line should be changed to:  for line_num,line in enumerate(loglines):

File: Detectors/FIT/benchmark/README.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+e.g. if you wish to monitor 50 pp (pythia) events with Geant3 as the VMC backend using the FIT detector and utilizing parallel mode with 2 simulation workers AND keep track of FairMQ devices, you can do:
+
+
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/README.md]: This line should be deleted

File: Detectors/FIT/benchmark/README.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+---
+
+Similarly you can monitor the digitization routine as: 
+
+`$> ./monitor.sh o2-sim-digitizer-workflow -b --run | tee  o2digi.log`
+
+NB! notice the `--run` that is needed (only digitization) in order to overload the PIPE `|` command in DPL (Data Processing Layer).
+
+
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/README.md]: This line should be deleted

**Comment by mslupeck** [comment]
> Thanks. Note that these tools are based on already existing scripts used in various production tests. Probably it would be useful to add new features directly there.

Do you have a specific production test in mind, that could be extended?

File: Detectors/FIT/benchmark/process.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+            # save the first 100 lines in o2digi.log
+            loglines = [next(logfile) for line in range(100)]
+
+        # declare list for PID numbers and names
+        pid_vals = []
+        pid_names = []
+
+        # loop through lines to find PIDs
+        for line_num,line in enumerate(loglines)
+            pid_line = re.findall('Starting (\\w.+) on pid (\\d+)', loglines[line])
```
Comment (zsaldic) [inline Detectors/FIT/benchmark/process.py]: Line 135:` pid_line = re.findall('Starting (\\w.+) on pid (\\d+)', loglines[line])` has to be changed to 
`pid_line = re.findall('Starting (\\w.+) on pid (\\d+)',line)`

Comment (mslupeck) [reply Detectors/FIT/benchmark/process.py]: Done

**Pull request merged**